File: 20/day20.rb
def presents1(house)
  total = 0
  1.upto(house).each { |n| total += n * 10 if house % n == 0 }
  total
end
raise unless presents1(1) == 10
raise unless presents1(4) == 70

def presents2(house)
  total = 0
  min = [house / 50, 1].max
  min.upto(house) { |n| total += n * 11 if house % n == 0 }
  total
end
raise unless presents2(1) == 11

min_presents = 36000000
n = (1..Float::INFINITY).detect do |house|
  next unless house % 360 == 0 # semi-cheat
  p = presents1(house)
  puts "at #{house}, #{p}" if house % 10000 == 0
  p >= min_presents
end
puts "Part 1: #{n}"
n2 = (n..Float::INFINITY).detect do |house|
  next unless house % 360 == 0
  p = presents2(house)
  puts "at #{house}, #{p}" if house % 10000 == 0
  p >= min_presents
end
puts "Part 2: #{n2}"

File: 20/day20.c
#include<stdio.h>
#include<stdlib.h>

int main() {
  int min_presents = 36000000;
  int max_houses = 1000000;
  int houses[max_houses];
  for (int elf = 1; elf < max_houses; elf += 1) {
    for (int house = elf; house < max_houses; house += elf) {
      houses[house] += elf * 10;
    }
  }
  for (int i = 0; i < max_houses; i += 1) {
    if (houses[i] >= min_presents) {
      printf("Part 1: %d\n", i);
      break;
    }
  }

  int houses2[max_houses];
  for (int elf = 1; elf < max_houses; elf += 1) {
    for (int house = elf; (house < max_houses && house <= 50 * elf); house += elf) {
      houses2[house] += elf * 11;
    }
  }
  for (int i = 0; i < max_houses; i += 1) {
    if (houses2[i] >= min_presents) {
      printf("Part 2: %d\n", i);
      break;
    }
  }
}
